### PR TITLE
Support double quotation in Path and optimize the Path String usage v2

### PR DESCRIPTION
**Pull request by Genius-pig**
Two goals:
(1) To support double quoted nodes in any level , e.g., root."s.g"."device.1"."sensor.1" : root -> "s.g" -> "device.1" -> "sensor.1"
(2) In antlr, a path has already been split to nodes. For example, root.sg1.d1.s1 to [root, sg1, d1, s1]. It‘s no need to generate a Path and split again in metadata, so this PR's goal is to reduce split actions in the metadata.

### write speed code:
```java
package org.apache.iotdb.db.integration;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.Statement;
import org.apache.iotdb.jdbc.IoTDBSQLException;
import org.junit.Test;

public class IoTDBBenchMark {

  private static void hello()
      throws Exception {
    Class.forName("org.apache.iotdb.jdbc.IoTDBDriver");
    try (Connection connection = DriverManager.getConnection("jdbc:iotdb://127.0.0.1:6667/", "root", "root");
        Statement statement = connection.createStatement()) {
      try {
        statement.execute("SET STORAGE GROUP TO root.hello.world.boris.jack.lucy.trump.clinton.good.sg1");
        statement.execute(
            "CREATE TIMESERIES root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1.s1 WITH DATATYPE=INT64, ENCODING=RLE, COMPRESSOR=SNAPPY");
        statement.execute(
            "CREATE TIMESERIES root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1.s2 WITH DATATYPE=INT64, ENCODING=RLE, COMPRESSOR=SNAPPY");
        statement.execute(
            "CREATE TIMESERIES root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1.s3 WITH DATATYPE=INT64, ENCODING=RLE, COMPRESSOR=SNAPPY");
      } catch (IoTDBSQLException e) {
        System.out.println(e.getMessage());
      }

      for(int j = 0; j <= 200; j++) {
        for (int i = 0; i <= 100000; i++) {
          statement.addBatch(
              "insert into root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1(timestamp, s1, s2, s3) values("
                  + j * 100000 + i + "," + 1 + "," + 1 + "," + 1 + ")");
        }
        statement.executeBatch();
        statement.clearBatch();
      }

      ResultSet resultSet = statement.executeQuery("select * from root where time <= 10");
      if (resultSet != null) {
        System.out.println("--------------------------");
        final ResultSetMetaData metaData = resultSet.getMetaData();
        final int columnCount = metaData.getColumnCount();
        for (int i = 0; i < columnCount; i++) {
          System.out.print(metaData.getColumnLabel(i + 1) + " ");
        }
        System.out.println();
        while (resultSet.next()) {
          for (int i = 1; ; i++) {
            System.out.print(resultSet.getString(i));
            if (i < columnCount) {
              System.out.print(", ");
            } else {
              System.out.println();
              break;
            }
          }
        }
        System.out.println("--------------------------\n");
      }
    }
  }


  @Test
  public void hellok() throws Exception {
    long start = System.currentTimeMillis();
    hello();
    long end = System.currentTimeMillis();
    System.out.println(end - start);
  }

}

```

### pr and master write cost
pr | master
---|---
561021ms | 786882ms
571697ms | 819016ms

### master write GC activity
image

### pr write GC activity
image

### master write total memory

image

### pr write total memory

image

In a word, this PR's write speed, GC activity, and memory are much better than master.

### query speed code

```
package org.apache.iotdb.db.integration;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.Statement;
import org.apache.iotdb.jdbc.IoTDBSQLException;
import org.junit.Test;

public class IoTDBBenchMark {

  private static void hello()
      throws Exception {
    Class.forName("org.apache.iotdb.jdbc.IoTDBDriver");
    try (Connection connection = DriverManager.getConnection("jdbc:iotdb://127.0.0.1:6667/", "root", "root");
        Statement statement = connection.createStatement()) {
      try {
        statement.execute("SET STORAGE GROUP TO root.hello.world.boris.jack.lucy.trump.clinton.good.sg1");
        statement.execute(
            "CREATE TIMESERIES root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1.s1 WITH DATATYPE=INT64, ENCODING=RLE, COMPRESSOR=SNAPPY");
        statement.execute(
            "CREATE TIMESERIES root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1.s2 WITH DATATYPE=INT64, ENCODING=RLE, COMPRESSOR=SNAPPY");
        statement.execute(
            "CREATE TIMESERIES root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1.s3 WITH DATATYPE=INT64, ENCODING=RLE, COMPRESSOR=SNAPPY");
      } catch (IoTDBSQLException e) {
        System.out.println(e.getMessage());
      }

      for (int i = 0; i <= 100000; i++) {
        statement.addBatch(
            "insert into root.hello.world.boris.jack.lucy.trump.clinton.good.sg1.d1(timestamp, s1, s2, s3) values("
                + i + "," + 1 + "," + 1 + "," + 1 + ")");
      }
      statement.executeBatch();
      statement.clearBatch();

      ResultSet resultSet = null;
      ResultSet resultSet1 = null;
      for(int i = 0; i < 700000; i++) {
        resultSet = statement.executeQuery("select last * from root");
      }

      if (resultSet1 != null) {
        System.out.println("--------------------------");
        final ResultSetMetaData metaData = resultSet1.getMetaData();
        final int columnCount = metaData.getColumnCount();
        for (int i = 0; i < columnCount; i++) {
          System.out.print(metaData.getColumnLabel(i + 1) + " ");
        }
        System.out.println();
        while (resultSet1.next()) {
          for (int i = 1; ; i++) {
            System.out.print(resultSet1.getString(i));
            if (i < columnCount) {
              System.out.print(", ");
            } else {
              System.out.println();
              break;
            }
          }
        }
        System.out.println("--------------------------\n");
      }

      if (resultSet != null) {
        System.out.println("--------------------------");
        final ResultSetMetaData metaData = resultSet.getMetaData();
        final int columnCount = metaData.getColumnCount();
        for (int i = 0; i < columnCount; i++) {
          System.out.print(metaData.getColumnLabel(i + 1) + " ");
        }
        System.out.println();
        while (resultSet.next()) {
          for (int i = 1; ; i++) {
            System.out.print(resultSet.getString(i));
            if (i < columnCount) {
              System.out.print(", ");
            } else {
              System.out.println();
              break;
            }
          }
        }
        System.out.println("--------------------------\n");
      }
    }

  }


  @Test
  public void hellok() throws Exception {
    long start = System.currentTimeMillis();
    hello();
    long end = System.currentTimeMillis();
    System.out.println(end - start);
  }

}
```

### pr query total memory
image

### master query total memory
image

### pr query gc activity
image

### master query gc activity
image

### pr and master session write cost

pr | master
---|---
101473ms | 107207ms
102302ms | 116278ms

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/metadata/MTree.java [hunk 1]
```diff
@@ -671,11 +668,11 @@ private int getCountInGivenLevel(MNode node, int targetLevel) {
    *
    * <p>result: [name, alias, storage group, dataType, encoding, compression, offset]
    */
-  List<String[]> getAllMeasurementSchema(ShowTimeSeriesPlan plan) throws MetadataException {
-    List<String[]> res;
-    String[] nodes = MetaUtils.getNodeNames(plan.getPath().getFullPath());
+  List<Pair<PartialPath, String[]>> getAllMeasurementSchema(ShowTimeSeriesPlan plan) throws MetadataException {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MTree.java:671]: recover this, leave this to another PR， The pair structure is not good also

File: server/src/main/java/org/apache/iotdb/db/qp/logical/crud/InOperator.java [hunk 2]
```diff
@@ -55,7 +56,7 @@
    * @param path path
    * @param values values
    */
-  public InOperator(int tokenIntType, Path path, boolean not, Set<String> values) {
+  public InOperator(int tokenIntType, PartialPath path, boolean not, Set<String> values) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/qp/logical/crud/InOperator.java:59]: If the PartialPath extends the Path, this is no need to change.

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 3]
```diff
@@ -445,8 +446,8 @@ public void delete(String deviceId, String measurementId, long startTime, long e
    */
   public QueryDataSource query(SingleSeriesExpression seriesExpression, QueryContext context,
       QueryFileManager filePathsManager)
-      throws StorageEngineException, QueryProcessException {
-    String deviceId = seriesExpression.getSeriesPath().getDevice();
+      throws StorageEngineException, QueryProcessException, IllegalPathException {
+    PartialPath deviceId = new PartialPath(seriesExpression.getSeriesPath().getDevice());
```
[the remark this answers is not in the document]
Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/metadata/PartialPath.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  public boolean startsWith(String[] otherNodes) {
+    for(int i = 0; i < otherNodes.length; i++) {
+      if(!nodes[i].equals(otherNodes[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public Path toTSFilePath() {
```
[the remark this answers is not in the document]
Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/metadata/PartialPath.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/metadata/MTree.java (hunk 1 above)
Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/metadata/MTree.java:671]: OK

File: server/src/main/java/org/apache/iotdb/db/qp/logical/crud/InOperator.java (hunk 2 above)
Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/qp/logical/crud/InOperator.java:59]: fixed

File: server/src/main/java/org/apache/iotdb/db/metadata/PartialPath.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+     }
+    return true;
+  }
+
+  @Override
+  public int hashCode() {
+    return this.toString().hashCode();
+  }
+
+  public String getLastNode() {
```
[the remark this answers is not in the document]
Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/metadata/PartialPath.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/metadata/PartialPath.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    return nodes[nodes.length - 1];
+  }
+
+  public String getFirstNode() {
+    return nodes[0];
+  }
+
+  public String getPathWithoutLastNode() {
```
[the remark this answers is not in the document]
Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/metadata/PartialPath.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/qp/logical/crud/InOperator.java [hunk 7]
```diff
@@ -129,7 +130,7 @@ public void reverseFunc() {
                   .endsWith("\"")) ? new Binary(val.substring(1, val.length() - 1))
                   : new Binary(val));
         }
-        ret = In.getUnaryExpression(singlePath, binaryValues, not);
+        ret = In.getUnaryExpression(new Path(singlePath.getDevice(), singlePath.getMeasurement()), binaryValues, not);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/qp/logical/crud/InOperator.java:133]: No need to construct a Path

Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/qp/logical/crud/InOperator.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 8]
```diff
@@ -108,7 +108,7 @@ public void insert(InsertRowPlan insertRowPlan) {
       Object value = insertRowPlan.getValues()[i];
       memSize += MemUtils.getRecordSize(insertRowPlan.getSchemas()[i].getType(), value);
 
-      write(insertRowPlan.getDeviceId(), insertRowPlan.getMeasurements()[i],
+      write(insertRowPlan.getDeviceId().getFullPath(), insertRowPlan.getMeasurements()[i],
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java:111]: I think we can use PartialPath directly here. There is no need to transfer it to String. If so, you need to change the type of `memTableMap` in `AbstractMemTable` to `Map<PartialPath, Map<String, IWritableMemChunk>> memTableMap`.
And by the way, i think you should rewrite the both hashCode and compareTo methods in PartialPath. We can compare each string in string array`nodes` instead of calling the fullPath() method which will retain it in memory until the PartialPath is cleaned by GC.

File: server/src/main/java/org/apache/iotdb/db/engine/merge/task/MergeFileTask.java [hunk 9]
```diff
@@ -151,7 +152,10 @@ private void moveMergedToOld(TsFileResource seqFile) throws IOException {
       TsFileIOWriter oldFileWriter = getOldFileWriter(seqFile);
 
       // filter the chunks that have been merged
-      oldFileWriter.filterChunks(context.getUnmergedChunkStartTimes().get(seqFile));
+      oldFileWriter.filterChunks(
+          context.getUnmergedChunkStartTimes().get(seqFile).entrySet().stream()
+              .collect(Collectors.toMap(Entry::getKey, Map.Entry::getValue))
+      );
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/engine/merge/task/MergeFileTask.java:158]: Why here we need to copy it?

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 10]
```diff
@@ -736,7 +736,7 @@ public void insert(InsertRowPlan insertRowPlan) throws WriteProcessException {
       // insert to sequence or unSequence file
       insertToTsFileProcessor(insertRowPlan,
           insertRowPlan.getTime() > partitionLatestFlushedTimeForEachDevice.get(timePartitionId)
-              .getOrDefault(insertRowPlan.getDeviceId(), Long.MIN_VALUE));
+              .getOrDefault(insertRowPlan.getDeviceId().getFullPath(), Long.MIN_VALUE));
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:739]: Same as above, I think we can use PartialPath directly.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 11]
```diff
@@ -216,11 +216,11 @@ public void insert(InsertRowPlan insertRowPlan) throws WriteProcessException {
     }
 
     // update start time of this memtable
-    tsFileResource.updateStartTime(insertRowPlan.getDeviceId(), insertRowPlan.getTime());
+    tsFileResource.updateStartTime(insertRowPlan.getDeviceId().getFullPath(), insertRowPlan.getTime());
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:219]: Same as above, if we change the key's type of the map to PartialPath, we don't have to call the getFullPath() method.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Review by Alima777 — APPROVED** [review]

File: server/src/main/java/org/apache/iotdb/db/engine/merge/task/MergeFileTask.java (hunk 9 above)
Comment (Genius-pig) [reply server/src/main/java/org/apache/iotdb/db/engine/merge/task/MergeFileTask.java]: fixed

**Review by qiaojialin — APPROVED** [review]
Better to change "." to TsFileConstant.PATH_SEPARATOR
